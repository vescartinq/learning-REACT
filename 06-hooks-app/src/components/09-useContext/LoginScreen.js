import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const LoginScreen = () => {
  const { setUser } = useContext(UserContext);

  let user = {
    id: 123,
    name: 'Victor',
  };

  return (
    <div>
      <h1>LoginScreen</h1>
      <hr />
      <button className="btn btn-primary" onClick={() => setUser(user)}>
        Login
      </button>
    </div>
  );
};
