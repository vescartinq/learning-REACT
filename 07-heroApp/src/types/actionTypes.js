export const actionTypes = {
  LOGIN: '[auth] login',
  LOGOUT: '[auth] logout',
};
