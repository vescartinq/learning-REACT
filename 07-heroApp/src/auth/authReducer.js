import { actionTypes } from '../types/actionTypes';

// const state = {
//     name: 'Victor',
//     logged: true
// }

export const authReducer = (state = {}, action) => {
  switch (action.type) {
    case actionTypes.LOGIN:
      return {
        ...action.payload,
        logged: true,
      };

    case actionTypes.LOGOUT:
      return {
        logged: false,
      };

    default:
      return state;
  }
};
