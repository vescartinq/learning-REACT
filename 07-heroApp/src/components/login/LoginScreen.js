import React, { useContext } from 'react';
import { AuthContext } from '../../auth/AuthContext';
import { actionTypes } from '../../types/actionTypes';

export const LoginScreen = ({ history }) => {
  const { dispatch } = useContext(AuthContext);

  const handleLogin = () => {
    // recupera la última ubicación despues de loggout -> lastPath debe ser igual que el indicado en PrivateRoute.js
    const lastPath = localStorage.getItem('lastPath') || '/';

    dispatch({
      type: actionTypes.LOGIN,
      payload: {
        name: 'Victor',
      },
    });

    history.replace(lastPath);
    // history.push('/'); // redirecciona a la dirección que indiquemos
    // history.replace('/'); // no visita el login, va directo al Dashboard
    // Lo utilizamos en esta App para redirect a /marvel
  };

  return (
    <div className="container mt-5">
      <h1>Login</h1>
      <hr />
      <button className="btn btn-primary" onClick={handleLogin}>
        Login
      </button>
    </div>
  );
};
