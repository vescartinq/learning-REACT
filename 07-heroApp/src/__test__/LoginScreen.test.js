import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { LoginScreen } from '../components/login/LoginScreen';
import { AuthContext } from '../auth/AuthContext';
import { types } from '../types/types';

describe('LoginScreen', () => {
  const history = {
    replace: jest.fn(),
  };

  const contextValue = {
    dispatch: jest.fn(),
    user: {
      logged: false,
    },
  };

  const wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      <LoginScreen history={history} />
    </AuthContext.Provider>
  );

  test('should render <LoginScreen/> component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should execute dispatch and navigation', () => {
    const handleClick = wrapper.find('button').prop('onClick');
    handleClick();

    // Dispara el dispatch para hacer LogIn
    expect(contextValue.dispatch).toHaveBeenCalledWith({
      type: types.login,
      payload: {
        name: 'Victor',
      },
    });

    // Redirecciona al hacer click
    expect(history.replace).toHaveBeenCalledWith('/');

    // Redirecciona a la página almacenada como última visita
    localStorage.setItem('lastPath', '/dc');
    handleClick();
    expect(history.replace).toHaveBeenCalledWith('/dc');
  });
});
