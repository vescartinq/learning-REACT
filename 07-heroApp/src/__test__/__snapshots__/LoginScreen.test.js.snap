// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen should render <LoginScreen/> component 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
